oraclepaas/resource_mysql_service_instance.go

```diff
@@ -0,0 +1,662 @@
+		log.Printf("[Error] : Error while extracting MySQL component information from TF file. : %s", err)
+		return err
+	}
+
+	log.Printf("[DEBUG] : Testing Create :%v : ", &input)
+
+	_, err = client.CreateServiceInstance(&input)
+
+	if err != nil {
+		log.Printf("[Error] : Error while creating MySQL Service Instance : %v", err)
+		return err
+	}
+
+	d.SetId(input.ServiceParameters.ServiceName)
+	return resourceOraclePAASMySQLServiceInstanceUpdate(d, meta)
+}
+
+/**
+getServiceParameters gets the values from the terraform resource file, and updates the inputParameter
+with the respective values for calling the "Create"
+*/
+func getServiceParameters(d *schema.ResourceData) (mysql.ServiceParameters, error) {
+
+	input := &mysql.ServiceParameters{
+		ServiceName:       d.Get("service_name").(string),
+		BackupDestination: d.Get("backup_destination").(string),
+		VMPublicKeyText:   d.Get("ssh_public_key").(string),
+	}
+
+	if value, ok := d.GetOk("availability_domain"); ok {
+		input.AvailabilityDomain = value.(string)
+	}
+
+	if value, ok := d.GetOk("metering_frequency"); ok {
+		input.MeteringFrequency = value.(string)
+	}
+
+	if value, ok := d.GetOk("region"); ok {
+		input.Region = value.(string)
+	}
+
+	if value, ok := d.GetOk("service_description"); ok {
+		input.ServiceDescription = value.(string)
+	}
+
+	err := expandCloudStorage(d, input)
+	if err != nil {
+		return *input, err
+	}
+
+	return *input, nil
+}
+
+/**
+Expands and reads the values in the Cloud_Storage list specified in the terraform file.
+*/
+func expandCloudStorage(d *schema.ResourceData, parameter *mysql.ServiceParameters) error {
+
+	cloudStorageInfo := d.Get("cloud_storage_configuration").([]interface{})
+
+	if parameter.BackupDestination == string(mysql.ServiceInstanceBackupDestinationBoth) || parameter.BackupDestination == string(mysql.ServiceInstanceBackupDestinationOSS) {
+		if len(cloudStorageInfo) == 0 {
+			return fmt.Errorf("`cloud_storage_configuration` must be set if `backup_destination` is set to `OSS` or `BOTH`")
+		}
+	}
+
+	if len(cloudStorageInfo) > 0 {
+		attrs := cloudStorageInfo[0].(map[string]interface{})
+		parameter.CloudStorageContainer = attrs["cloud_storage_container"].(string)
+		parameter.CloudStorageContainerAutoGenerate = attrs["create_if_missing"].(bool)
+		if val, ok := attrs["cloud_storage_username"].(string); ok && val != "" {
+			parameter.CloudStorageUsername = val
+		}
+		if val, ok := attrs["cloud_storage_password"].(string); ok && val != "" {
+			parameter.CloudStoragePassword = val
+		}
+	}
+	return nil
+}
+
+func expandEM(input map[string]interface{}, parameter *mysql.MySQLParameters) error {
+
+	emInfo := input["enterprise_monitor_configuration"].([]interface{})
+
+	log.Printf("[DEBUG] input                       : %v", input)
+	log.Printf("[DEBUG] parameter.EnterpriseMonitor : %v", parameter.EnterpriseMonitor)
+	log.Printf("[DEBUG] emInfo                      : %d", len(emInfo))
+
+	if parameter.EnterpriseMonitor == "Yes" {
```

Comment:
Can we assume that a user will want `enterprise_monitor` set to yes if they have the `enterprise_monitor_configuration` block specified. If so, we can just set it before it's sent off to the sdk instead of checking that the user sets it.

Reply:
Done.